content/blog/2021/08/19/WritingFastConstraintStreamsSecretRecipe.adoc

```diff
@@ -0,0 +1,119 @@
+	end
+----
+
+As you can see, this is a lot of DRL which fundamentally does this:
+
+1. Infer the first shift in a sequence of consecutive shifts.
+2. Infer the last shift in a sequence of consecutive shifts.
+3. Infer all the non-overlapping sequences.
+4. Penalize sequences longer than what is prescribed by the contract.
+
+Now consider how the same constraint is accomplished with Constraint Streams, using the experimental consecutive constraint collector:
+
+[source, java]
+----
+    Constraint consecutiveWorkingDays(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(MinMaxContractLine.class)
+            .filter(minMaxContractLine -> minMaxContractLine
+                .getContractLineType() == ContractLineType.CONSECUTIVE_WORKING_DAYS &&
+                minMaxContractLine.isEnabled())
+            .join(ShiftAssignment.class,
+                Joiners.equal(ContractLine::getContract, ShiftAssignment::getContract))
+            .groupBy((contract, shift) -> shift.getEmployee(),
+                (contract, shift) -> contract,
+                ExperimentalConstraintCollectors.consecutive((contract, shift) -> shift.getShiftDate(),
+                    ShiftDate::getDayIndex))
+            .flattenLast(ConsecutiveInfo::getConsecutiveSequences)
+            .filter((employee, contract, shiftList) -> contract.isViolated(shiftList.getLength()))
+            .penalize("consecutiveWorkingDays", HardSoftScore.ONE_SOFT,
+                    (employee, contract, shiftList) -> contract.getViolationAmount(shiftList.getLength()));
+    }
+----
+
+This constraint uses the experimental consecutive constraint collector to get us a list of all sequences of consecutive shifts.
+This list is then https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/index.html#constraintStreamsFlattening[flattened], giving us each sequence individually.
+As you can see, this is a much more concise implementation of the same behavior, with the brunt of the logic hidden inside the constraint collector itself.
+
+This constraint collector https://github.com/kiegroup/optaplanner/blob/8.10.0.Final/optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/ExperimentalConstraintCollectors.java[is not part of our public API] and we consider it experimental.
+Before we make it part of the public API, we are looking for your feedback to make sure it fits the various use cases that are out there.
+If you have any questions or see issues in applying this pattern to your own constraints, please reach out to us on https://kie.zulipchat.com/#narrow/stream/232679-optaplanner[KIE Zulip], or https://stackoverflow.com/questions/tagged/optaplanner[ask a question on StackOverflow].
+
+== Conclusion
+
+We will continue improving Constraint Streams as we find more and more problems to solve.
+If you want to make sure we can solve your problems too, share them with us.
```

Comment:
Link to optaplanner "get help" page?

Ps: In general, it's good to have a link to optaplanner.org front page in each blog post, for SEO in case the blog post get copied to other outlets.